extensions/amp-form/0.1/amp-form.js

```diff
@@ -176,9 +188,75 @@ export class AmpForm {
 }
 
 
+function reportValidity(state) {
+  reportFormValidity(state.form);
+}
+
+/**
+ * Reports validity for the first invalid input - if any.
+ * @param {!HTMLFormElement} form
+ * @private
+ */
+function reportFormValidity(form) {
+  const inputs = toArray(form.querySelectorAll('input,select,textarea'));
+  for (let i = 0; i < inputs.length; i++) {
+    if (!inputs[i].checkValidity()) {
+      reportInputValidity(inputs[i]);
+      break;
+    }
+  }
+}
+
+
+/**
+ * Revalidates the currently focused input after a change.
+ * @param {!KeyboardEvent} event
+ * @private
+ */
+function onInvalidInputKeyUp_(event) {
+  validationBubble.hide();
+  if (!event.target.checkValidity()) {
+    validationBubble.show(event.target, event.target.validationMessage);
+  }
+}
+
+
+/**
+ * Hides validation bubble and removes listeners on the invalid input.
+ * @param {!Event} event
+ * @private
+ */
+function onInvalidInputBlur_(event) {
+  validationBubble.hide();
+  event.target.removeEventListener('blur', onInvalidInputBlur_);
+  event.target.removeEventListener('keyup', onInvalidInputKeyUp_);
+}
+
+
+/**
+ * Focuses and reports the invalid message of the input in a message bubble.
+ * @param {!HTMLInputElement} input
+ * @private
+ */
+function reportInputValidity(input) {
+  input./*OK*/focus();
+
+  // Remove any previous listeners on the same input. This avoids adding many
+  // listeners on the same element when the user submit pressing Enter or any
+  // other method to submit the form without the element losing focus.
+  input.removeEventListener('blur', onInvalidInputBlur_);
+  input.removeEventListener('keyup', onInvalidInputKeyUp_);
+
+  input.addEventListener('keyup', onInvalidInputKeyUp_);
+  input.addEventListener('blur', onInvalidInputBlur_);
+  validationBubble.show(input, input.validationMessage);
+}
+
+
 /**
  * Installs submission handler on all forms in the document.
  * @param {!Window} win
+ * @private
```

Comment:
+1

On Fri, Jun 24, 2016 at 9:44 AM, Justin Ridgewell <notifications@github.com>
wrote:

> In extensions/amp-form/0.1/amp-form.js
> <https://github.com/ampproject/amphtml/pull/3734#discussion_r68425990>:
>
> >  /**
> >   * Installs submission handler on all forms in the document.
> >   * @param {!Window} win
> > + * @private
>
> I don't think there's a need to mark un-exported functions as private.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/3734/files/957c812439bb3654460089220b746a2cff48ca96#r68425990>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAFeT7uaY0EY2nyCT6InRZN1wVEQrVAoks5qPAl1gaJpZM4I8WLq>
> .
>